interop/xds/server/server.go

```diff
@@ -78,8 +93,89 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
 }
 
 func (s *testServiceImpl) UnaryCall(ctx context.Context, in *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+	response := &testpb.SimpleResponse{ServerId: s.serverID, Hostname: s.hostname}
+
+behavior:
+	for _, headerVal := range getRPCBehaviorMetadata(ctx) {
+		switch {
+		//If the value matches "sleep-<int>", the server should wait the specified number of seconds before resuming behavior matching and RPC processing.
+		case strings.HasPrefix(headerVal, sleepPfx):
+			{
+				sleep, err := strconv.Atoi(headerVal[len(sleepPfx):])
+				if err != nil {
+					return nil, status.Errorf(codes.InvalidArgument, "invalid format for rpc-behavior header %v, must be '%v<int>' instead", headerVal, sleepPfx)
+				}
+
+				time.Sleep(time.Duration(sleep) * time.Second)
+			}
+		// If the value matches "keep-open", the server should never respond to the request and behavior matching ends.
+		case strings.HasPrefix(headerVal, keepOpenVal):
+			{
+				<-ctx.Done()
+			}
+		// If the value matches "error-code-<int>", the server should respond with the specified status code and behavior matching ends.
+		case strings.HasPrefix(headerVal, errorCodePfx):
+			{
+				code, err := strconv.Atoi(headerVal[len(errorCodePfx):])
+				if err != nil {
+					return nil, status.Errorf(codes.InvalidArgument, "invalid format for rpc-behavior header %v, must be '%v<int>' instead", headerVal, errorCodePfx)
+				}
+
+				return nil, status.Errorf(codes.Code(code), "rpc failed as per the rpc-behavior header value: %v", headerVal)
+			}
+		// If the value matches "success-on-retry-attempt-<int>", and the value of the "grpc-previous-rpc-attempts" metadata field is equal
+		// to the specified number, the normal RPC processing should resume and behavior matching ends.
+		case strings.HasPrefix(headerVal, successOnRetryPfx):
+			{
+				wantRetry, err := strconv.Atoi(headerVal[len(successOnRetryPfx):])
+				if err != nil {
+					return nil, status.Errorf(codes.InvalidArgument, "invalid format for rpc-behavior header %v, must be '%v<int>' instead", headerVal, successOnRetryPfx)
+				}
+
+				mdRetry := getMetadataValues(ctx, rpcAttemptsMdKey)
+				curRetry, err := strconv.Atoi(mdRetry[0])
+				if err != nil {
+					return nil, status.Errorf(codes.InvalidArgument, "invalid format for grpc-previous-rpc-attempts header: %v", mdRetry[0])
+				}
+
+				if curRetry == wantRetry {
+					break behavior
+				}
+			}
+		// A value can have a prefix "hostname=<string>" followed by a space. In that case, the rest of the value should only be applied
+		// if the specified hostname matches the server's hostname.
```

Comment:
Based on my reading of https://github.com/grpc/grpc-java/blob/master/interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java#L320, it looks like we can't use a `switch` after all.

So, we need to handle cases like `hostname=foo sleep-5`, and we need to process the `sleep-5` only if our server's hostname is `foo`.

So, this loop basically has to be:
```golang
for <iterate through the list of rpc behavior metadata values> {
  // currVal stores current value
  if hostname prefix {
    if hostname does not match { continue }
    // update currVal and run through the rest of the matchers
  }
  if sleep prefix { ... }
  if call behavior success after number retries prefix { ... }
  if keep open prefix { ... }
  if error prefix { ... }
}
```

Apologies for the back and forth. But hopefully, we've figured out how this is supposed to work this time around :)

Reply:
pushed a commit with the above change

Reply:
Fair enough.